expr/tagquery/expression.go

```diff
@@ -45,134 +37,336 @@ func (e Expressions) Strings() []string {
-	return e.Key == other.Key && e.Operator == other.Operator && e.Value == other.Value
-}
+	if effectiveOperator == MATCH || effectiveOperator == MATCH_TAG || effectiveOperator == NOT_MATCH {
+		if len(resCommon.value) > 0 && resCommon.value[0] != '^' {
+			resCommon.value = "^(?:" + resCommon.value + ")"
+		}
 
-func (e *Expression) StringIntoBuilder(builder *strings.Builder) {
-	if e.Operator == MATCH_TAG || e.Operator == PREFIX_TAG {
-		builder.WriteString("__tag")
+		// no need to run regular expressions that match any string
+		// so we update the operator to MATCH_ALL/NONE
+		if resCommon.value == "^(?:.*)" || resCommon.value == "^.*" || resCommon.value == "^(.*)" {
+			switch effectiveOperator {
+			case MATCH:
+				return &expressionMatchAll{expressionCommon: resCommon, originalOperator: originalOperator}, nil
+			case MATCH_TAG:
+				return &expressionMatchAll{expressionCommon: resCommon, originalOperator: originalOperator}, nil
+			case NOT_MATCH:
+				return &expressionMatchNone{expressionCommon: resCommon, originalOperator: originalOperator}, nil
+			}
+		}
+
+		valueRe, err := regexp.Compile(resCommon.value)
+		if err != nil {
+			return nil, err
+		}
+
+		// check for special case when regular expression matches
+		// empty value and update operator accordingly
+		matchesEmpty := valueRe.MatchString("")
+
+		switch effectiveOperator {
+		case MATCH:
+			return &expressionMatch{expressionCommonRe: expressionCommonRe{expressionCommon: resCommon, valueRe: valueRe, matchesEmpty: matchesEmpty}}, nil
+		case NOT_MATCH:
+			return &expressionNotMatch{expressionCommonRe: expressionCommonRe{expressionCommon: resCommon, valueRe: valueRe, matchesEmpty: matchesEmpty}}, nil
+		case MATCH_TAG:
+			if matchesEmpty {
+				return nil, fmt.Errorf(invalidExpressionError, expr)
+			}
+			return &expressionMatchTag{expressionCommonRe: expressionCommonRe{expressionCommon: resCommon, valueRe: valueRe, matchesEmpty: matchesEmpty}}, nil
+		}
 	} else {
-		builder.WriteString(e.Key)
+		switch effectiveOperator {
+		case EQUAL:
+			return &expressionEqual{expressionCommon: resCommon}, nil
+		case NOT_EQUAL:
+			return &expressionNotEqual{expressionCommon: resCommon}, nil
+		case PREFIX:
+			return &expressionPrefix{expressionCommon: resCommon}, nil
+		case HAS_TAG:
+			return &expressionHasTag{expressionCommon: resCommon}, nil
+		case NOT_HAS_TAG:
+			return &expressionNotHasTag{expressionCommon: resCommon}, nil
+		case PREFIX_TAG:
+			return &expressionPrefixTag{expressionCommon: resCommon}, nil
+		case MATCH_ALL:
+			return &expressionMatchAll{expressionCommon: resCommon, originalOperator: originalOperator}, nil
+		case MATCH_NONE:
+			return &expressionMatchNone{expressionCommon: resCommon, originalOperator: originalOperator}, nil
+		}
 	}
-	e.Operator.StringIntoBuilder(builder)
-	builder.WriteString(e.Value)
+
+	return nil, fmt.Errorf("ParseExpression: Invalid operator in expression %s", expr)
 }
 
+func ExpressionsAreEqual(expr1, expr2 Expression) bool {
```

Comment:
why not make an Expression.Equals method?

Reply:
Agree, that's better